File: android/java/org/chromium/chrome/browser/BraveFeatureUtil.java
/* Copyright (c) 2022 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.build.annotations.NullMarked;

@JNINamespace("chrome::android")
@NullMarked
public abstract class BraveFeatureUtil {
    // The method sets a feature state that is passed in `enabled` var. However it can
    // set the feature to a Default state if Default state is what is `enabled` var.
    // For example: A feature `dummy_name` has 3 states and a Default value is Disabled:
    // 0: Default (Disabled)
    // 1: Enabled
    // 2: Disabled
    // enableFeature("dummy_name", false, true) sets the `dummy_name` feature state to Default
    // enableFeature("dummy_name", false, false) sets the `dummy_name` feature state to Disabled
    public static void enableFeature(
            String featureName, boolean enabled, boolean fallbackToDefault) {
        BraveFeatureUtilJni.get().enableFeature(featureName, enabled, fallbackToDefault);
    }

    @NativeMethods
    interface Natives {
        void enableFeature(String featureName, boolean enabled, boolean fallbackToDefault);
    }
}

File: browser/android/brave_feature_util.cc
/* Copyright (c) 2020 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include <string>

#include "base/android/jni_string.h"
#include "base/check.h"
#include "base/feature_list.h"
#include "base/strings/string_number_conversions.h"
#include "chrome/android/chrome_jni_headers/BraveFeatureUtil_jni.h"
#include "chrome/browser/about_flags.h"
#include "chrome/browser/browser_process.h"
#include "components/webui/flags/feature_entry.h"
#include "components/webui/flags/pref_service_flags_storage.h"

namespace chrome {
namespace android {

int GetNumberOfOptions(const std::string& internal_name) {
  DCHECK(about_flags::GetCurrentFlagsState());
  if (!about_flags::GetCurrentFlagsState()) {
    return 0;
  }
  const flags_ui::FeatureEntry* entry =
      about_flags::GetCurrentFlagsState()->FindFeatureEntryByName(
          internal_name);
  DCHECK(entry);
  if (!entry) {
    return 0;
  }
  return entry->NumOptions();
}

static void JNI_BraveFeatureUtil_EnableFeature(
    JNIEnv* env,
    const base::android::JavaRef<jstring>& featureName,
    jboolean enabled,
    jboolean fallback_to_default) {
  std::string feature_name =
      base::android::ConvertJavaStringToUTF8(env, featureName);
  std::string disabled_value =
      fallback_to_default
          ? "0"
          : base::NumberToString(GetNumberOfOptions(feature_name) - 1);
  enabled ? feature_name += "@1" : feature_name += "@" + disabled_value;
  flags_ui::PrefServiceFlagsStorage flags_storage(
      g_browser_process->local_state());
  about_flags::SetFeatureEntryEnabled(&flags_storage, feature_name, true);
}

}  // namespace android
}  // namespace chrome

DEFINE_JNI(BraveFeatureUtil)
